### PR TITLE
Feature/wrap sfm data

### PR DESCRIPTION
**Pull request by swarrier246**
I've written a python unittest for the functions I've used from the SfmData and SfmTrack wrapper. Will add tests for other methods after I've had some doubts clarified.
Changed "measurements" in SfmTrack to Measurements, to maintain uniformity in how vectors are named (specializations.h). That change was also reflected in all other files using track.measurements

### CODE REVIEW

File: gtsam/slam/dataset.h
Comment (akshay-krishnan) [inline gtsam/slam/dataset.h:223]: I think this should be renamed to `measurements`. The convention is to use small letters to start member variable names.

File: gtsam/gtsam.i
Comment (akshay-krishnan) [inline gtsam/gtsam.i:2766]: I think the MATLAB wrapper also needs them to have basic methods to do anything with them. Like at() or push_back() for vectors. That is the case for other vectors that have been wrapped in this file.

File: gtsam/gtsam.i
Comment (akshay-krishnan) [inline gtsam/gtsam.i:2774]: Do the existing versions of these methods not work? Is there a reason for the change?

File: gtsam/slam/dataset.h
Comment (akshay-krishnan) [inline gtsam/slam/dataset.h:262]: Could we use lower case `measurements` for the member name here as well?

Comment (akshay-krishnan) [reply gtsam/slam/dataset.h:262]: Could you also please document what these measurements are? I understand the one in SfmTrack, but what does the one in SfmData represent?

File: gtsam/gtsam.i
Comment (akshay-krishnan) [inline gtsam/gtsam.i:2766]: I dont think we need to wrap SfmCamera, this is being wrapped as PinholeCameraCal3Bundler in another PR, and there are a few methods in other classes that use it. Please refer to https://github.com/borglab/gtsam/pull/556. Perhaps we can merge that PR if it is needed for this code.

File: python/gtsam/specializations.h
Comment (akshay-krishnan) [inline python/gtsam/specializations.h:18]: Do we need vector cameras? Maybe we could call them VectorPinholeCameraCal3Bundler after that PR has been merged. Or maybe in that PR itself, if we need them.

**Review by akshay-krishnan — CHANGES_REQUESTED** [review]

File: gtsam/slam/dataset.h
Comment (akshay-krishnan) [inline gtsam/slam/dataset.h:285]: I think this should be a const reference, like `void add_camera(const SfmCamera& cam)`.

File: gtsam/slam/dataset.h
Comment (akshay-krishnan) [inline gtsam/slam/dataset.h:281]: Please change the argument to a const reference, unless there's a good reason not to.

File: gtsam/slam/dataset.h
Comment (akshay-krishnan) [inline gtsam/slam/dataset.h:246]: Please use const reference arguments.

**Review by akshay-krishnan — CHANGES_REQUESTED** [review]

File: gtsam/gtsam.i
Comment (swarrier246) [reply gtsam/gtsam.i:2766]: i haven't implemented the matlab wrapper yet. These are dummy classes

File: gtsam/slam/dataset.h
Comment (swarrier246) [reply gtsam/slam/dataset.h:262]: it's the same: measurements within the tracks. But I'll update that in documentation

File: python/gtsam/preamble.h
Comment (johnwlambert) [inline python/gtsam/preamble.h:13]: why do we need to add these here? Didn't catch what `PYBIND11_MAKE_OPAQUE` does for us

File: gtsam/gtsam.i
Comment (johnwlambert) [reply gtsam/gtsam.i:2774]: i propose we come up with a better name than `siftIndex` while we are at it here, since we won't to be using non-SIFT detections, descriptions, and correspondences very soon

Comment (swarrier246) [reply gtsam/gtsam.i:2774]: do we need them at all? Because to my understanding, SiftIndex seems to be the feature coordinates at a particular measurement - then why is SfmMeasurement a pair of <size_t, Point2>? Why not <size_t, SiftIndex>?

File: python/gtsam/preamble.h
Comment (swarrier246) [reply python/gtsam/preamble.h:13]: This explains it well: https://pybind11.readthedocs.io/en/stable/advanced/cast/stl.html

File: gtsam/slam/dataset.h
Comment (akshay-krishnan) [inline gtsam/slam/dataset.h:250]: Could you point me to where this method is being used? I'm not sure why it is returned by reference.

Comment (swarrier246) [reply gtsam/slam/dataset.h:250]: Nowhere, actually. It was originally intended to return all measurements in a track(I wanted to avoid using a for loop to get measurements, but it returns an SfmMeasurement object, which we'd need to wrap to get the elements of, so I decided not to use it)

File: python/CMakeLists.txt
Comment (akshay-krishnan) [inline python/CMakeLists.txt:41]: All three of these are not being wrapped. I think they can be removed.

File: python/gtsam/preamble.h
Comment (akshay-krishnan) [inline python/gtsam/preamble.h:13]: Remove commented lines?

File: python/gtsam/specializations.h
Comment (akshay-krishnan) [inline python/gtsam/specializations.h:16]: Remove commented lines?

File: python/gtsam/tests/test_SfmData.py
Comment (akshay-krishnan) [inline python/gtsam/tests/test_SfmData.py:36]: Is the randomness of any use here? You might as well pick random numbers yourself. It helps to have determinism in tests.

File: python/gtsam/tests/test_SfmData.py
Comment (akshay-krishnan) [inline python/gtsam/tests/test_SfmData.py:63]: I don't think the randomness here is particularly useful. It doesn't hurt much either so its up to you. Also, point2 can be double, so why use randint?

File: python/gtsam/tests/test_SfmData.py
Comment (akshay-krishnan) [inline python/gtsam/tests/test_SfmData.py:9]: Ive not seen this kind of separation in author names before. Totally fine if it has been done elsewhere. Otherwise, I think you can list names in the order of the contribution to this file?

**Review by akshay-krishnan — CHANGES_REQUESTED** [review]
Looking much cleaner now! Leaving some final comments.

File: python/gtsam/tests/test_SfmData.py
Comment (swarrier246) [reply python/gtsam/tests/test_SfmData.py:9]: I think I based this off of John's PR #556 . Let me know if you still want me to change it

File: python/gtsam/tests/test_SfmData.py
Comment (swarrier246) [reply python/gtsam/tests/test_SfmData.py:63]: No particular reason, just seemed simpler. I'll add arbitrary numbers instead of the np.random at all places

File: python/gtsam/tests/test_SfmData.py
Comment (akshay-krishnan) [reply python/gtsam/tests/test_SfmData.py:9]: No, its up to you.

File: python/gtsam/tests/test_SfmData.py
Comment (akshay-krishnan) [inline python/gtsam/tests/test_SfmData.py:60]: Commented code, can be removed.

File: python/gtsam/tests/test_SfmData.py
Comment (akshay-krishnan) [inline python/gtsam/tests/test_SfmData.py:20]: Commented code, can be removed.

File: python/gtsam/tests/test_SfmData.py
Comment (akshay-krishnan) [inline python/gtsam/tests/test_SfmData.py:21]: I am not sure if this is standard practice for importing in Python. Hasn't gtsam already been imported above?

File: python/gtsam/preamble.h
Comment (akshay-krishnan) [inline python/gtsam/preamble.h:12]: Did you change any spacing or remove a new line here? Undoing that would remove this file from the PR.

File: gtsam/gtsam.i
Comment (akshay-krishnan) [inline gtsam/gtsam.i:2784]: It looks like the filename argument is a `const std::string& ` in dataset.h

**Review by akshay-krishnan — CHANGES_REQUESTED** [review]

File: python/gtsam/tests/test_SfmData.py
Comment (swarrier246) [reply python/gtsam/tests/test_SfmData.py:21]: Yeah, other tests have done it too (refer https://github.com/borglab/gtsam/blob/develop/python/gtsam/tests/test_Point2.py)

File: gtsam/gtsam.i
Comment (swarrier246) [reply gtsam/gtsam.i:2784]: that is repeated across a lot of functions(eg readBal, load2D, etc). I'm not sure why it is being used the way it is. Maybe @ProfFan could shed some light on it?

Comment (ProfFan) [reply gtsam/gtsam.i:2784]: I think this is just history :)

Comment (dellaert) [reply gtsam/gtsam.i:2784]: const and & are mostly ignored in .i files. Just using string is fine.

File: gtsam/gtsam.i
Comment (dellaert) [inline gtsam/gtsam.i:2767]: I'd rather see a new constructor than an imperative "set". "add" is fine as it is "part of the construction" of the object.

File: gtsam/slam/dataset.h
Comment (dellaert) [inline gtsam/slam/dataset.h:230]: see my comment from .i file.

File: gtsam/slam/dataset.h
Comment (dellaert) [inline gtsam/slam/dataset.h:241]: how about returning a const& ?

File: gtsam/slam/dataset.h
Comment (dellaert) [inline gtsam/slam/dataset.h:247]: can you use emplace?

**Review by dellaert — COMMENTED** [review]
I'll just comment, as Akshay requested changes, he can approve after you make changes.

File: python/gtsam/tests/test_SfmData.py
Comment (johnwlambert) [inline python/gtsam/tests/test_SfmData.py:76]: is the measurement returned as a Numpy array, rather than as a typed SfmMeasurement object? is that preferred?

Comment (johnwlambert) [reply python/gtsam/tests/test_SfmData.py:76]: Discussed Pair vs. NamedTuple, but ship has sailed, resolved.

Comment (akshay-krishnan) [reply python/gtsam/tests/test_SfmData.py:76]: Sorry I missed the discussion @johnwlambert . Do you think it would more readable if the tuple was unpacked locally like `cam_idx, point = tracks.measurements(0)`? Might save some effort on creating a class for the measurements and changing existing code in dataset.h.

File: gtsam/gtsam.i
Comment (akshay-krishnan) [inline gtsam/gtsam.i:2765]: About this new constructor, what happens to the measurements if you use them to triangulate a point? Should the constructor not accept measurements as an argument as well?

File: python/gtsam/preamble.h
Comment (akshay-krishnan) [reply python/gtsam/preamble.h:12]: This is not resolved yet.

**Review by akshay-krishnan — CHANGES_REQUESTED** [review]

File: gtsam/gtsam.i
Comment (swarrier246) [reply gtsam/gtsam.i:2765]: you mean a vector of measurements as a constructor argument?

Comment (dellaert) [reply gtsam/gtsam.i:2765]: No, it's acceptable to add them one at a time.

**Review by akshay-krishnan — APPROVED** [review]
Nice work! Thanks for repeatedly making the requested changes.

**Comment by ProfFan** [comment]
Mac builds are broken by upstream boost issue https://github.com/boostorg/serialization/issues/219

**Pull request merged**
